packages/core/src/types.ts

```diff
@@ -0,0 +1,370 @@
+	 */
+	validateLanguageOptions(languageOptions: LanguageOptions): void;
+
+	/**
+	 * Helper for esquery that allows languages to match nodes against
+	 * class. esquery currently has classes like `function` that will
+	 * match all the various function nodes. This method allows languages
+	 * to implement similar shorthands.
+	 */
+	matchesSelectorClass?(
+		className: string,
+		node: SyntaxElement,
+		ancestry: Array<SyntaxElement>,
+	): boolean;
+
+	/**
+	 * Parses the given file input into its component parts. This file should not
+	 * throws errors for parsing errors but rather should return any parsing
+	 * errors as parse of the ParseResult object.
+	 */
+	parse(file: File, context: LanguageContext): ParseResult; // Future: | Promise<ParseResult>;
+
+	/**
+	 * Creates SourceCode object that ESLint uses to work with a file.
+	 */
+	createSourceCode(
+		file: File,
+		input: ParseResult,
+		context: LanguageContext,
+	): SourceCode; // Future: | Promise<SourceCode>;
+}
+
+/**
+ * Plugin-defined options for the language.
+ */
+export type LanguageOptions = Record<string, unknown>;
+
+/**
+ * The context object that is passed to the language plugin methods.
+ */
+export interface LanguageContext {
+	languageOptions: LanguageOptions;
+}
+
+/**
+ * Represents a file read by ESLint.
+ */
+export interface File {
+	/**
+	 * The path that ESLint uses for this file. May be a virtual path
+	 * if it was returned by a processor.
+	 */
+	path: string;
+
+	/**
+	 * The path to the file on disk. This always maps directly to a file
+	 * regardless of whether it was returned from a processor.
+	 */
+	physicalPath: string;
+
+	/**
+	 * Indicates if the original source contained a byte-order marker.
+	 * ESLint strips the BOM from the `body`, but this info is needed
+	 * to correctly apply autofixing.
+	 */
+	bom: boolean;
+
+	/**
+	 * The body of the file to parse.
+	 */
+	body: string | Uint8Array;
+}
+
+/**
+ * Represents the successful result of parsing a file.
+ */
+export interface OkParseResult {
+	/**
+	 * Indicates if the parse was successful. If true, the parse was successful
+	 * and ESLint should continue on to create a SourceCode object and run rules;
+	 * if false, ESLint should just report the error(s) without doing anything
+	 * else.
+	 */
+	ok: true;
+
+	/**
+	 * The abstract syntax tree created by the parser. (only when ok: true)
+	 */
+	ast: SyntaxElement;
+
+	/**
+	 * Any additional data that the parser wants to provide.
+	 */
+	[key: string]: any;
+}
+
+/**
+ * Represents the unsuccessful result of parsing a file.
+ */
+export interface NotOkParseResult {
+	/**
+	 * Indicates if the parse was successful. If true, the parse was successful
+	 * and ESLint should continue on to create a SourceCode object and run rules;
+	 * if false, ESLint should just report the error(s) without doing anything
+	 * else.
+	 */
+	ok: false;
+
+	/**
+	 * Any parsing errors, whether fatal or not. (only when ok: false)
+	 */
+	errors?: Array<FileError>;
```

Comment:
```suggestion
	errors: Array<FileError>;
```

I think this should be mandatory with `ok:false`.

Reply:
And Linter indeed expects this property:

https://github.com/eslint/eslint/blob/7c78ad9d9f896354d557f24e2d37710cf79a27bf/lib/linter/linter.js#L939